In [1]:
import sys
sys.path.append('/Users/camillecu/Downloads/KUL/master_thesis/master_thesis_github')

In [2]:

# import pysparnn.cluster_index as ci
# import scipy.sparse
# import os
import math
import numpy as np
import pandas as pd
import random
import scipy.sparse as sp
from utils import train_test_split, df_to_matrix ,matrix_to_df_2, threshold_interactions_df, matrix_to_df,set_intersection,get_0_and_p_index,set_diff, matrix_to_full_df, threshold_interactions_df_plus, train_test_split_csr

!pip install surprise
from surprise import Reader, accuracy
from surprise import SVD
from surprise import Dataset
from collections import Counter
from collections import defaultdict

%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt

from pct.tree.heuristic.Heuristic import Heuristic5
from pct.tree.heuristic.NumericHeuristic_pair import NumericHeuristic5
from pct.tree.splitter.splitter_yahoo import Splitter
import pct.tree.splitter.splitter_yahoo as splitter_yahoo

from pct.tree.Yahootree import Tree


[notice] A new release of pip is available: 25.0.1 -> 25.1.1
[notice] To update, run: pip install --upgrade pip


In [3]:
# import df_pairwise from csv
df_pairwise = pd.read_csv('/Users/camillecu/Downloads/KUL/master_thesis/master_thesis_github/yahoo_data/df_pairwise.csv')
df_pairwise_subset = df_pairwise.sample(n=2000, random_state=42)

# Pivot to create user-item rating matrix
rating_matrix = df_pairwise_subset.pivot(index='user_id',columns='item_id',values='rating').fillna(0)

# Create item_type mapping (for splitter logic)
item_type_map = df_pairwise_subset.set_index('item_id')['item_type'].to_dict()
# Assign it to the splitter_yahoo module's variable
splitter_yahoo.item_type_map = item_type_map


In [4]:
tree_strategy1 = Tree(min_instances=15, max_depth=3)
tree_strategy1.fit(rating_matrix, df_pairwise_subset['rating'], target_weights=None, strategy=1)

✅Initializing Splitter...
✅ Calling build()...
🔄 Building tree at depth 0 with 1090 users and 2000 items.
[DEBUG] select_pair called with strategy=1

[DEBUG] Selected itemA = 160961 with error = 23.557, type = artist
🔍 Pair found: 160961, 157933
🔍 Found 6 users who rated either 160961 or 157933.
User IDs in all_subset_users: [68784, 180677, 102088, 223530, 110028, 166589]
🔍 Found 1084 users with no ratings for both items.
🔍 User ratings for 160961: {110028: 1.0, 166589: 0.1, 180677: 0.1, 223530: 0.1}
🔍 User ratings for 157933: {68784: 1.0, 102088: 0.1}
Group A: 4 users, Group B: 2 users, Unknown: 1084 users
🔄 Building tree at depth 1 with 4 users and 4 items.
🍃 Not enough users (4) to split (min_instances=15). Making leaf.
🔄 Building tree at depth 1 with 2 users and 2 items.
🍃 Not enough users (2) to split (min_instances=15). Making leaf.
🔄 Building tree at depth 1 with 1084 users and 1084 items.
[DEBUG] select_pair called with strategy=1

[DEBUG] Selected itemA = 9771 with error = 21.

In [5]:
tree_strategy2 = Tree(min_instances=15, max_depth=3)
tree_strategy2.fit(rating_matrix, df_pairwise_subset['rating'], target_weights=None, strategy=2)

✅Initializing Splitter...
✅ Calling build()...
🔄 Building tree at depth 0 with 1090 users and 2000 items.
[DEBUG] select_pair called with strategy=2

[DEBUG] Selected itemA = 160961 with error = 23.557, type = artist

[DEBUG] itemA = 160961, vecA = [0 0 0 ... 0 0 0]
  Compared to itemB = 157933, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 161583, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 9771, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 22935, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 23717, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 41507, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 71521, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 122972, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 153616, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 212293, vecB = [0 0 0 ... 0 0 0], similari

In [6]:
tree_strategy3 = Tree(min_instances=15, max_depth=3)
tree_strategy3.fit(rating_matrix, df_pairwise_subset['rating'], target_weights=None, strategy=3)

✅Initializing Splitter...
✅ Calling build()...
🔄 Building tree at depth 0 with 1090 users and 2000 items.
[DEBUG] select_pair called with strategy=3

[DEBUG] Selected itemA = 160961 with error = 23.557, type = artist

[DEBUG] itemA = 160961, vecA = [0 0 0 ... 0 0 0]
  Compared to itemB = 157933, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 161583, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 9771, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 22935, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 23717, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 41507, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 71521, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 122972, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 153616, vecB = [0 0 0 ... 0 0 0], similarity = 0.000
  Compared to itemB = 212293, vecB = [0 0 0 ... 0 0 0], similari

In [ ]:
# build tree structure
tree_strategy1.print_tree_structure()

In [3]:

my_seed = 7
random.seed(my_seed)
np.random.seed(my_seed)

def load_user_item_matrix(filepath):
    data = []

    with open(filepath, 'r') as f:
        while True:
            user_line = f.readline()
            if not user_line:
                break

            user_line = user_line.strip()
            if '|' not in user_line:
                continue

            user_id, num_ratings = user_line.split('|')
            user_id = int(user_id)
            num_ratings = int(num_ratings)

            for _ in range(num_ratings):
                item_line = f.readline().strip()
                parts = item_line.split('\t')
                if len(parts) < 2:
                    continue
                item_id, score = parts[0], parts[1]
                data.append([user_id, int(item_id), int(score)])

    df = pd.DataFrame(data, columns=['user_id', 'item_id', 'rating'])
    return df


In [4]:
def safe_int(x):
    return int(x) if x != 'None' else 0


def load_track_data(filepath):
    track_data = []
    with open(filepath, 'r') as f:
        for line in f:
            parts = line.strip().split('|')
            track_id = safe_int(parts[0])
            album_id = safe_int(parts[1])
            artist_id = safe_int(parts[2])
            genre_ids = [safe_int(g) for g in parts[3:]] if len(parts) > 3 else []
            track_data.append([track_id, album_id, artist_id, genre_ids])
    return pd.DataFrame(track_data, columns=['track_id', 'album_id', 'artist_id', 'genre_ids'])

def load_album_data(filepath):
    album_data = []
    with open(filepath, 'r') as f:
        for line in f:
            parts = line.strip().split('|')
            album_id = safe_int(parts[0])
            artist_id = safe_int(parts[1])
            genre_ids = [safe_int(g) for g in parts[2:]] if len(parts) > 2 else []
            album_data.append([album_id, artist_id, genre_ids])
    return pd.DataFrame(album_data, columns=['album_id', 'artist_id', 'genre_ids'])


# read artistData2
def load_artist_data(filepath):
    with open(filepath, 'r') as f:
        artist_ids = [int(line.strip()) for line in f]
    return pd.DataFrame(artist_ids, columns=['artist_id'])

# read genreData2
def load_genre_data(filepath):
    with open(filepath, 'r') as f:
        genre_ids = [int(line.strip()) for line in f]
    return pd.DataFrame(genre_ids, columns=['genre_id'])

In [5]:
track_df = load_track_data('/Users/camillecu/Downloads/KUL/master_thesis/master_thesis_github/yahoo_data/ydata-ymusic-kddcup-2011-track2/trackData2.txt')
track_df

track_id  album_id  artist_id                 genre_ids
0              1    106710     281667  [214765, 162234, 155788]
1              2    280977     233685   [131552, 173467, 48505]
2              3     38422     219136    [61215, 201738, 88853]
3              4    119529     166863            [17453, 35389]
4              5     16742     294690    [61215, 34486, 274088]
...          ...       ...        ...                       ...
224036    296100    166516      33011   [274088, 199606, 88853]
224037    296101         0          0                        []
224038    296102    153644     289056  [158282, 139095, 242383]
224039    296105     68336       6613                   [82064]
224040    296110    281202     140302                   [61215]

[224041 rows x 4 columns]

In [6]:
album_df = load_album_data('/Users/camillecu/Downloads/KUL/master_thesis/master_thesis_github/yahoo_data/ydata-ymusic-kddcup-2011-track2/albumData2.txt')
album_df

album_id  artist_id                genre_ids
0             0          0                 [214765]
1             6     228091  [158282, 81520, 242383]
2            19      85028                 [103715]
3            30      16832                  [31567]
4            33      26330         [149962, 209270]
...         ...        ...                      ...
52824    296104     125866                 [158282]
52825    296106     221637           [116130, 9432]
52826    296107          0                  [61215]
52827    296108      93277                  [61215]
52828    296109      84885                 [287681]

[52829 rows x 3 columns]

In [7]:
artist_df = load_artist_data('/Users/camillecu/Downloads/KUL/master_thesis/master_thesis_github/yahoo_data/ydata-ymusic-kddcup-2011-track2/artistData2.txt')
artist_df

artist_id
0             16
1             23
2             35
3             40
4             49
...          ...
18669     295987
18670     296007
18671     296012
18672     296013
18673     296041

[18674 rows x 1 columns]

In [8]:
genre_df = load_genre_data('/Users/camillecu/Downloads/KUL/master_thesis/master_thesis_github/yahoo_data/ydata-ymusic-kddcup-2011-track2/genreData2.txt')
genre_df

genre_id
0         208
1         315
2         642
3        1075
4        1271
..        ...
562    292093
563    293670
564    293688
565    294138
566    294310

[567 rows x 1 columns]

In [9]:
train_path = '/Users/camillecu/Downloads/KUL/master_thesis/master_thesis_github/yahoo_data/ydata-ymusic-kddcup-2011-track2/trainIdx2.txt'
my_seed = 7
random.seed(my_seed)
np.random.seed(my_seed)
train_df = load_user_item_matrix(train_path)


In [10]:
train_df

user_id  item_id  rating
0               0    28341      90
1               0    51210      90
2               0    79500      90
3               0    82317      90
4               0    98399      90
...           ...      ...     ...
61944401   249011   270557      90
61944402   249011   273574      90
61944403   249011   286938      90
61944404   249011   287681      80
61944405   249011   295381      90

[61944406 rows x 3 columns]

In [11]:
duplicate_counts = train_df.groupby(['user_id', 'item_id']).size()
duplicates = duplicate_counts[duplicate_counts > 1]

print(f"🔁 Number of duplicated (user_id, item_id) pairs: {len(duplicates)}")
print(duplicates.head())


🔁 Number of duplicated (user_id, item_id) pairs: 0
Series([], dtype: int64)


In [12]:
print(train_df['item_id'].max())      # highest item_id
print(track_df['track_id'].max())     # highest track_id
print(train_df['item_id'].isin(track_df['track_id']).mean())  # proportion to match

print(train_df['item_id'].isin(album_df['album_id']).mean())

print(train_df['item_id'].isin(artist_df['artist_id']).mean())

print(train_df['item_id'].isin(genre_df['genre_id']).mean())



296110
296110
0.4385845107627636
0.1925648621120041
0.3114063600835885
0.05744426704164376


In [13]:
def build_item_metadata(track_df, album_df, artist_df, genre_df):
    rows = []

    # 1. Tracks
    for _, row in track_df.iterrows():
        rows.append({
            'item_id': row['track_id'],
            'item_type': 'track',
            'artist_id': row['artist_id'],
            'genre_ids': row['genre_ids']
        })

    # 2. Albums
    for _, row in album_df.iterrows():
        rows.append({
            'item_id': row['album_id'],
            'item_type': 'album',
            'artist_id': row['artist_id'],
            'genre_ids': row['genre_ids']
        })

    # 3. Artists
    for artist_id in artist_df['artist_id']:
        rows.append({
            'item_id': artist_id,
            'item_type': 'artist',
            'artist_id': artist_id,
            'genre_ids': 0
        })

    # 4. Genres
    for genre_id in genre_df['genre_id']:
        rows.append({
            'item_id': genre_id,
            'item_type': 'genre',
            'artist_id': 0,
            'genre_ids': [genre_id]
        })

    return pd.DataFrame(rows)

In [14]:
item_meta_df = build_item_metadata(track_df, album_df, artist_df, genre_df)

train_with_meta = train_df.merge(item_meta_df, on='item_id', how='inner')

# When loading the data, sort by a stable column (e.g., user_id)
train_with_meta = train_with_meta.sort_values('user_id').reset_index(drop=True)

train_with_meta

user_id  item_id  rating item_type  artist_id genre_ids
0               0    28341      90    artist      28341         0
1               0   232604      30    artist     232604         0
2               0   237425      50    artist     237425         0
3               0   245257      50    artist     245257         0
4               0    23301      50    artist      23301         0
...           ...      ...     ...       ...        ...       ...
61944401   249011   141075      90    artist     141075         0
61944402   249011   150385      80     genre          0  [150385]
61944403   249011   152077      90    artist     152077         0
61944404   249011   158282      80     genre          0  [158282]
61944405   249011   295381      90    artist     295381         0

[61944406 rows x 6 columns]

In [15]:
df = train_with_meta.copy()

# Filter: keep only artist & genre (drop album, track)
df = df[~df['item_type'].isin(['album', 'track'])].copy()

print(f"🎯 Sample shape: {df.shape}")
print(f"👤 Users: {df['user_id'].nunique()}, 🎵 Items: {df['item_id'].nunique()}")

🎯 Sample shape: (22848233, 6)
👤 Users: 248344, 🎵 Items: 19241


In [16]:
filtered_df = threshold_interactions_df(df,'user_id','item_id',100,500)

Starting interactions info
Number of rows: 248344
Number of cols: 19241
Sparsity: 0.478%
Ending interactions info
Number of rows: 4644
Number of columns: 8938
Sparsity: 10.674%


In [17]:
# filtered_df = threshold_interactions_df_plus(df, min_items_per_user=500, min_artists_per_user=20, min_genres_per_user=10,min_users_per_item=150)
my_seed = 7
random.seed(my_seed)
np.random.seed(my_seed)
sampled_users = np.random.choice(filtered_df['user_id'].unique(), size=2000, replace=False)
sampled_df = filtered_df[filtered_df['user_id'].isin(sampled_users)]
print(f"🎯 Sample shape: {sampled_df.shape}")
print(f"👤 Users: {sampled_df['user_id'].nunique()}, 🎵 Items: {sampled_df['item_id'].nunique()}")
filtered_df = threshold_interactions_df(sampled_df,'user_id','item_id',100,500)

🎯 Sample shape: (1941009, 6)
👤 Users: 2000, 🎵 Items: 8938
Starting interactions info
Number of rows: 2000
Number of cols: 8938
Sparsity: 10.858%
Ending interactions info
Number of rows: 1701
Number of columns: 5011
Sparsity: 17.890%


Sample shape: (1941009, 6)
👤 Users: 2000, 🎵 Items: 8938
Starting interactions info
Number of rows: 2000
Number of cols: 8938
Sparsity: 10.858%
Ending interactions info
Number of rows: 1701
Number of columns: 5011
Sparsity: 17.890%

In [18]:
filtered_df

user_id  item_id  rating item_type  artist_id genre_ids
1039            9   182744      30    artist     182744         0
1040            9   197011      50    artist     197011         0
1041            9   160820       0    artist     160820         0
1043            9    59402      30    artist      59402         0
1044            9   251901       0    artist     251901         0
...           ...      ...     ...       ...        ...       ...
61933948   248947   210727      50    artist     210727         0
61933956   248947   275456       0    artist     275456         0
61933959   248947   203741       0    artist     203741         0
61933965   248947   202602       0    artist     202602         0
61933966   248947   295880       0    artist     295880         0

[1524894 rows x 6 columns]

# Test Yahoo tree build

In [19]:
# convert ratings to 1 when rating > 50; if rating <= 50 but greater than 0, set to 0.1, otherwise set to 0
df_pairwise = filtered_df.copy()
df_pairwise['rating'] = df_pairwise['rating'].apply(lambda x: 1 if x > 50 else (0.1 if x > 0 else 0))


In [23]:
# drop genre_ids and artist_id columns
df_pairwise = df_pairwise.drop(columns=['genre_ids', 'artist_id'])

In [25]:
# export df_pairwise to csv
df_pairwise.to_csv('/Users/camillecu/Downloads/KUL/master_thesis/master_thesis_github/yahoo_data/df_pairwise.csv', index=False)
# convert df_pairwise to matrix

In [ ]:
# import df_pairwise from csv
df_pairwise = pd.read_csv('/Users/camillecu/Downloads/KUL/master_thesis/master_thesis_github/yahoo_data/df_pairwise.csv')
df_pairwise_subset = df_pairwise.sample(n=2000, random_state=42)


In [4]:
df_pairwise.head()

user_id  item_id  rating item_type
0        9   182744     0.1    artist
1        9   197011     0.1    artist
2        9   160820     0.0    artist
3        9    59402     0.1    artist
4        9   251901     0.0    artist

In [7]:
# type of value of rating column
print(df_pairwise['rating'].dtype)

float64


In [5]:
# Pivot to create user-item rating matrix
rating_matrix = df_pairwise_subset.pivot(
    index='user_id',
    columns='item_id',
    values='rating'
).fillna(0)

# Create item_type mapping (for splitter logic)
item_type_map = df_pairwise_subset.set_index('item_id')['item_type'].to_dict()


In [5]:
tree = Tree(min_instances=15)
tree.fit(df_pairwise_subset, df_pairwise_subset['rating'], target_weights=None, strategy=1)

✅ Numerical attributes: Index(['user_id', 'item_id', 'rating'], dtype='object')
✅ Categorical attributes: Index(['item_type'], dtype='object')
Initializing Splitter...
✅ Calling build()...


/Users/camillecu/Downloads/KUL/master_thesis/master_thesis_github/thesis_github/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)
  return reduction(axis=axis, out=out, **passkwargs)


TypeError: unsupported operand type(s) for ** or pow(): 'str' and 'int'

In [ ]:
all_user_ids = sorted(filtered_df['user_id'].unique())

def split_users_by_ratio(all_user_ids, ratio):
    n = len(all_user_ids)
    split_point = int(n * ratio)
    warm_users = all_user_ids[:split_point]
    cold_users = all_user_ids[split_point:]
    return warm_users, cold_users

# Example ratios from 10% to 50%
# ratios = [0.1, 0.2, 0.3, 0.4, 0.5]
# splits = {r: split_users_by_ratio(all_user_ids, r) for r in ratios}


## 10% warm

In [ ]:
warm_users_idx, cold_users_idx = split_users_by_ratio(all_user_ids, 0.1)

df_warm = filtered_df[filtered_df['user_id'].isin(warm_users_idx)].copy()
df_cold = filtered_df[filtered_df['user_id'].isin(cold_users_idx)].copy()

matrix_warm, rid_to_idx_warm, idx_to_rid_warm, cid_to_idx, idx_to_cid = df_to_matrix(
    df_warm, "user_id", "item_id", "rating")


matrix_cold, rid_to_idx_cold, idx_to_rid_cold, _, _ = df_to_matrix( 
    df_cold, "user_id", "item_id", "rating")



In [ ]:
def split_and_combine(strategy="artist-only"):
    """Handles both approaches with proper matrix alignment"""
    # Get full cold matrix and mappings
    matrix_cold, rid_to_idx_cold, _, cid_to_idx, _ = df_to_matrix(
        df_cold, "user_id", "item_id", "rating"
    )
    matrix_cold = matrix_cold.tocsr()

    # Create boolean masks
    artist_mask = np.isin(
        np.arange(matrix_cold.shape[1]), 
        [cid_to_idx[iid] for iid in df_cold[df_cold['item_type'] == 'artist']['item_id']]
    )
    genre_mask = ~artist_mask

    # Create aligned matrices
    matrix_cold_artist = matrix_cold.multiply(artist_mask)
    matrix_cold_genre = matrix_cold.multiply(genre_mask)
    matrix_cold_artist = matrix_cold.multiply(artist_mask).tocsr()
    matrix_cold_genre = matrix_cold.multiply(genre_mask).tocsr()

    al_artist, test_cold, _ = train_test_split(
        matrix_cold_artist, 
        split_count=30,
        fraction=None
    )
    
    if strategy == "artist-only":
        X_cold, K_cold, _ = train_test_split_csr(al_artist, 1)  
        return K_cold, X_cold, test_cold
    
    elif strategy == "hybrid":
        X_cold, K_cold, _ = train_test_split_csr(al_artist, 1)
        X_cold = X_cold + matrix_cold_genre
        return K_cold, X_cold, test_cold


In [ ]:
my_seed = 7
random.seed(my_seed)
np.random.seed(my_seed)

train_cold_K_artist, X_cold_artist, test_cold_artist  = split_and_combine("artist-only")
train_cold_K_hybrid, X_cold_hybrid, test_cold_hybrid = split_and_combine("hybrid")


In [ ]:
# shape of test_cold

print(f"Cold users in X: {len(np.unique(X_cold_artist.nonzero()[0]))}")
print(f"Cold items in X: {len(np.unique(X_cold_artist.nonzero()[1]))}")
print(f"test users in test: {len(np.unique(test_cold_artist.nonzero()[0]))}")
print(f"test items in test: {len(np.unique(test_cold_artist.nonzero()[1]))}")
print(f"train users in train: {len(np.unique(train_cold_K_artist.nonzero()[0]))}")
print(f"train items in train: {len(np.unique(train_cold_K_artist.nonzero()[1]))}")

print("-------------------")

# shape of test_cold
print(f"Shape of test_cold: {test_cold_artist.shape}")
print(f"Shape of train_cold: {train_cold_K_artist.shape}")
print(f"Shape of X_cold: {X_cold_artist.shape}")

Cold users in X: 1531
Cold items in X: 4793
test users in test: 1473
test items in test: 3890
train users in train: 665
train items in train: 576
-------------------
Shape of test_cold: (1531, 5011)
Shape of train_cold: (1531, 5011)
Shape of X_cold: (1531, 5011)
